src/raft.c

```diff
@@ -1304,13 +1310,40 @@ void callRaftPeriodic(RedisModuleCtx *ctx, void *arg)
         EntryCacheCompact(rr->logcache, rr->config.log_max_cache_size);
     }
 
-    /* Initiate snapshot if log size exceeds raft-log-file-max */
-    if (!rr->snapshot_in_progress && rr->config.log_max_file_size &&
-        raft_get_num_snapshottable_logs(rr->raft) > 0 &&
-        LogFileSize(rr->log) > rr->config.log_max_file_size) {
-        LOG_DEBUG("Raft log file size is %lu, initiating snapshot.",
-                  LogFileSize(rr->log));
-        initiateSnapshot(rr);
+    /* Initiate snapshot if log size exceeds raft-log-file-max
+     *
+     * Compaction happens in two steps:
+     * 1- We call LogCompactionBegin() and start writing entries to a new file.
+     * 2- Once we commit all the entries of the first file, we initiate the
+     *    snapshot.
+     */
+    if (!rr->snapshot_in_progress) {
+        bool start;
```

Comment:
The way I see, `redisraft.c` and `raft.c` are the owner of other objects. Moving logic into log makes things other way around and it becomes hard to test log implementation (as it is going to call `initiateSnapshot()` or `raft_get_commit_idx()` etc. Currently, `LogBlaBla()` functions just follows the orders. They don't reach to `rr->raft` or snapshot code. 

I think control should still be here but maybe we can improve things a bit (probably with better function names, e.g. you didn't like `LogFirstPageIdx()` (rightfully), maybe we need to come up with something else. 

Can we hide things without moving all the logic into log.c? (just thinking loud).

Reply:
The way we decided right now to implement the compaction should not affect raft.c. It will be simple if we will keep the encapsulation concept, and hide the two phases from raft.c, so once we decide to change the implementation, it will be easier.

Reply:
That's my point. `log.c` cannot hide it without knowing about snapshotting, periodic calls, libraft (rr->raft) etc. Moving these into log.c makes it too complex. Then, it'll be impossible to test `log.c` alone. 

What is your suggestion? Moving this "if check" into `log.c` completely or adding some other helper functions to `log.c`?

Reply:
From my perspective the whole logic here is `LogCompaction` -> during the log compaction we need information from raft obj, and can call to `initiateSnapshot` but it's a log operation, therefore the whole block here should move to a single func in log.c. This comment is not a blocker for merge, but it isn't a style comment, but a design one.

Reply:
`struct Log` is a dummy object, it gets orders from the owner and functions that use/mutate it are in log.c . Similar to `entrycache.c`, `file.c` or `nodeaddr.c`.  Even logImpl is too much for this file, should be in/around redisraft.c. I left there not to make PR larger. If we move extra logic into these files, there won't be any difference between `raft.c/redisraft.c` and these "dummy" files. 

I suggest following one scheme, owner knows all "dummy" things: `raft.c/redisraft.c` knows log, snapshot, shards, config, threads. "dummy" guys don't know each other. So, it means we control things from a single place. There is a clear boundary between files and objects. (Although we don't follow this rule in certain places)